pkg/virt-operator/resource/generate/components/prometheus.go

```diff
@@ -204,31 +252,30 @@ func NewPrometheusRuleSpec(ns string, workloadUpdatesEnabled bool) *v1.Prometheu
 					{
 						Record: "vec_by_virt_controllers_all_client_rest_requests_in_last_hour",
 						Expr: intstr.FromString(
-							fmt.Sprintf("sum by (pod) (sum_over_time(rest_client_requests_total{pod=~'virt-controller-.*', namespace='%s'}[60m]))", ns),
+							GeneratePromQlQueryForValueDifferenceInHourInterval("virt_controller_rest_client_requests"),
 						),
 					},
 					{
 						Record: "vec_by_virt_controllers_failed_client_rest_requests_in_last_hour",
 						Expr: intstr.FromString(
-							fmt.Sprintf("sum by (pod) (sum_over_time(rest_client_requests_total{pod=~'virt-controller-.*', namespace='%s', code=~'(4|5)[0-9][0-9]'}[60m]))", ns),
+							GeneratePromQlQueryForValueDifferenceInHourInterval("virt_controller_rest_client_requests_with_errors"),
 						),
 					},
 					{
 						Record: "vec_by_virt_controllers_all_client_rest_requests_in_last_5m",
 						Expr: intstr.FromString(
-							fmt.Sprintf("sum by (pod) (sum_over_time(rest_client_requests_total{pod=~'virt-controller-.*', namespace='%s'}[5m]))", ns),
+							GeneratePromQlQueryForValueDifferenceInFiveMinutesInterval("virt_controller_rest_client_requests"),
 						),
 					},
 					{
 						Record: "vec_by_virt_controllers_failed_client_rest_requests_in_last_5m",
 						Expr: intstr.FromString(
-							fmt.Sprintf("sum by (pod) (sum_over_time(rest_client_requests_total{pod=~'virt-controller-.*', namespace='%s', code=~'(4|5)[0-9][0-9]'}[5m]))", ns),
+							GeneratePromQlQueryForValueDifferenceInFiveMinutesInterval("virt_controller_rest_client_requests_with_errors"),
 						),
 					},
 					{
 						Alert: "VirtControllerRESTErrorsHigh",
 						Expr:  intstr.FromString("(vec_by_virt_controllers_failed_client_rest_requests_in_last_hour / vec_by_virt_controllers_all_client_rest_requests_in_last_hour) >= 0.05"),
-						For:   "5m",
```

Comment:
question: what happens if the for a minute we have very high quantity of REST calls which cause an alert to be triggered. What happens if after another 10 seconds all REST requests are back to normal and have very low failure rate. Does this mean that for 4 minutes we will see this alert firing over and over (since indeed in the last 5 minutes we had, on average, high amount of REST errors)?

If this is true then for 1 hour it'll be even worse - we would re-trigger the event over and over for an hour even if the problem is already solved.

I think that 5 minutes could be a bit extreme, but what about changing it to `3m` or something similar? Same goes for the hourly alerts of course.

Reply:
Hey Itamar,

if i understood correctly i think that firing is a state and that when we add `For:    "5m"` to the alert 
it means that Prometheus will fire an alert only if the condition (in the Expr's value) is true for 5m
so it would be a west of time to wait another 5 minutes when the error rate is high for at least one hour.
this is explained in here:

https://prometheus.io/docs/prometheus/latest/configuration/alerting_rules/

"The optional for clause causes Prometheus to wait for a certain duration between first encountering a new expression output vector element and counting an alert as firing for this element. In this case, **Prometheus will check that the alert continues to be active during each evaluation for 10 minutes before firing the alert. Elements that are active, but not firing yet, are in the pending state.**"

Reply:
Thanks for clarifying! :)